server/src/main/java/com/hedera/block/server/persistence/storage/write/BlockAsFileWriter.java

```diff
@@ -0,0 +1,111 @@
+     * get the {@link MetricsService}
+     * @param blockRemover valid, {@code non-null} instance of {@link BlockRemover} used to remove
+     * blocks in case of cleanup
+     * @param blockPathResolver valid, {@code non-null} instance of {@link BlockPathResolver} used
+     * to resolve paths to Blocks
+     */
+    BlockAsFileWriter(
+            @NonNull final BlockNodeContext blockNodeContext,
+            @NonNull final BlockRemover blockRemover,
+            @NonNull final BlockPathResolver blockPathResolver) {
+        this.metricsService = Objects.requireNonNull(blockNodeContext.metricsService());
+        this.blockRemover = Objects.requireNonNull(blockRemover);
+        this.blockPathResolver = Objects.requireNonNull(blockPathResolver);
+    }
+
+    @Override
+    public Optional<List<BlockItem>> write(@NonNull final List<BlockItem> toWrite) throws IOException {
+        if (toWrite.getFirst().hasBlockHeader()) {
+            curentBlock = Block.newBuilder().items(toWrite).build();
+        } else {
+            final List<BlockItem> currentItems = curentBlock.items();
+            currentItems.addAll(toWrite);
+            curentBlock = Block.newBuilder().items(currentItems).build();
+        }
+
+        if (toWrite.getLast().hasBlockProof()) {
+            metricsService.get(BlocksPersisted).increment();
+            return writeToFs(curentBlock);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    // todo we could recursively retry if exception occurs, then after a few attempts
+    // if we cannot persist, we must throw the initial exception
+    private Optional<List<BlockItem>> writeToFs(final Block blockToWrite) throws IOException {
+        final long number = blockToWrite.items().getFirst().blockHeader().number(); // fixme could be null, handle!
+
+        // todo we should handle cases where the block path exists, we do not expect it to
+        // exist at this stage, if it does, there is something wrong here
+        final Path blockToWritePathResolved = blockPathResolver.resolvePathToBlock(number);
+        Files.createDirectories(blockToWritePathResolved.getParent());
+        Files.createFile(blockToWritePathResolved);
+
+        try (final FileOutputStream fos = new FileOutputStream(blockToWritePathResolved.toFile())) {
+            Block.PROTOBUF.toBytes(blockToWrite).writeTo(fos);
+            // todo what should be fallback logic if something goes wrong here? we attempt to resolve the path
+            // with proper perms (is that necessary)? we must clean up and retry?
+        } catch (final IOException ioe) {
+            // todo handle properly
+            throw new UncheckedIOException(ioe);
+        } catch (final UncheckedIOException uioe) {
+            // todo handle properly
+            throw uioe;
+        } catch (final Exception e) {
+            // todo handle properly
+            throw new RuntimeException(e);
+        }
```

Comment:
Unless you have a _very_ good reason, do not catch `Exception`, ever (a vast array of bugs in real systems come from this seemingly innocuous block).
Instead, handle any specific exceptions you can, declare any checked exceptions necessary, and let callers handle things from there.